lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp

```diff
@@ -0,0 +1,345 @@
+  if (!pg_result) {
+    KATANA_LOG_FATAL("could not make property graph: {}", pg_result.error());
+  }
+  Graph graph = pg_result.value();
+
+  katana::gPrint(
+      "Read ", graph.num_nodes(), " nodes, ", graph.num_edges(), " edges\n");
+
+  if (startNode >= graph.size() || reportNode >= graph.size()) {
+    KATANA_LOG_ERROR(
+        "failed to set report: ", reportNode,
+        " or failed to set source: ", startNode, "\n");
+    assert(0);
+    abort();
+  }
+
+  auto it = graph.begin();
+  std::advance(it, startNode.getValue());
+  GNode source = *it;
+  it = graph.begin();
+  std::advance(it, reportNode.getValue());
+  GNode report = *it;
+
+  size_t approxNodeData = graph.size() * 64;
+  katana::Prealloc(1, approxNodeData);
+  katana::reportPageAlloc("MeminfoPre");
+
+  if (algo == deltaStep || algo == deltaTile) {
+    katana::gInfo("Using delta-step of ", (1 << stepShift), "\n");
+    KATANA_LOG_WARN(
+        "Performance varies considerably due to delta parameter.\n");
+    KATANA_LOG_WARN("Do not expect the default to be good for your graph.\n");
+  }
+
+  katana::do_all(katana::iterate(graph), [&graph](GNode n) {
+    graph.GetData<NodeMax>(n) = 0;
+    graph.GetData<NodeCount>(n) = 0;
+  });
+
+  katana::gInfo("Running ", ALGO_NAMES[algo], " algorithm\n");
+
+  katana::StatTimer execTime("Timer_0");
+  execTime.start();
+
+  katana::InsertBag<std::pair<uint32_t, Path*>> paths;
+  katana::InsertBag<Path*> path_pointers;
+
+  bool reachable = CheckReachability(&graph, source);
+
+  if (reachable) {
+    switch (algo) {
+    case deltaTile:
+      DeltaStepAlgo<SrcEdgeTile, OBIM>(
+          &graph, source, SrcEdgeTilePushWrap{&graph}, TileRangeFn(), &paths,
+          &path_pointers);
+      break;
+    case deltaStep:
+      DeltaStepAlgo<UpdateRequest, OBIM>(
+          &graph, source, ReqPushWrap(), OutEdgeRangeFn{&graph}, &paths,
+          &path_pointers);
+      break;
+    case deltaStepBarrier:
+      katana::gInfo("Using OBIM with barrier\n");
+      DeltaStepAlgo<UpdateRequest, OBIM_Barrier>(
+          &graph, source, ReqPushWrap(), OutEdgeRangeFn{&graph}, &paths,
+          &path_pointers);
+      break;
+
+    default:
+      std::abort();
+    }
+  }
+
+  execTime.stop();
+
+  if (reachable) {
+    std::multimap<uint32_t, Path*> paths_map;
+
+    for (auto pair : paths) {
+      paths_map.insert(std::make_pair(pair.first, pair.second));
+    }
+
+    katana::reportPageAlloc("MeminfoPost");
+
+    katana::gPrint("Node ", report, " has these k paths:\n");
+
+    uint32_t num = (paths_map.size() > numPaths) ? numPaths : paths_map.size();
+
+    auto it_report = paths_map.begin();
+
+    for (uint32_t iter = 0; iter < num; iter++) {
+      katana::gPrint(" ", report);
+      GNode parent = report;
+
+      const Path* path = it_report->second;
+      it_report++;
+
+      while (path->last != NULL) {
+        parent = path->parent;
+        katana::gPrint(" ", parent);
```

Comment:
nit: I think it's better to print a path from source to report node. Please write a recursive function to print the path in such a way.

Also please print the distance of the path.